components/brave_wallet/browser/bitcoin/bitcoin_knapsack_solver.cc

```diff
@@ -0,0 +1,168 @@
+    // Second pass(if no valid transactions from 1st pass): Forcedly pick
+    // yet not picked groups(strarting from largest).
+    for (int pass = 0; pass < 2; ++pass) {
+      if (has_valid_transaction_for_iteration) {
+        DCHECK_EQ(pass, 1);
+        break;
+      }
+
+      for (uint32_t group_index = 0; group_index < input_groups_.size();
+           ++group_index) {
+        bool pick_group =
+            pass == 0 ? base::RandInt(0, 1) : !picked_groups[group_index];
+        if (!pick_group) {
+          continue;
+        }
+
+        // TODO(apaymyshev): avoid copying transaction. Just keep track of
+        // current vbytes of transaction and optimize by cost of adding an
+        // input.
+        BitcoinTransaction next_transaction = cur_transaction.Clone();
+        next_transaction.AddInputs(input_groups_[group_index].inputs());
+
+        // Minimum fee required needed for this transaction to be accepted.
+        // Depends on transaction's size and current fee rate.
+        uint64_t min_fee = ApplyFeeRate(
+            fee_rate_,
+            BitcoinSerializer::CalcTransactionVBytes(next_transaction, true));
+
+        // Move everything except `min_fee` to change output(if any). Throw away
+        // possible transaction if resulting change amount is LT than dust
+        // threshold.
+        if (auto change_amount =
+                next_transaction.MoveSurplusFeeToChangeOutput(min_fee)) {
+          if (change_amount < dust_output_threshold) {
+            continue;
+          }
+        }
+
+        if (next_transaction.AmountsAreValid(min_fee)) {
+          has_valid_transaction_for_iteration = true;
+          // TODO(apaymyshev): Should we also add cost of spending change output
+          // in the future?
+          solutions.emplace(next_transaction.EffectiveFeeAmount(),
+                            next_transaction.Clone());
+
+          // Keep some best ones(less fee) in the container. Might be useful for
+          // logging later.
+          while (solutions.size() > 10) {
+            solutions.erase(std::prev(solutions.end()));
+          }
+        } else {
+          picked_groups[group_index] = true;
+          cur_transaction = std::move(next_transaction);
+        }
+      }
+    }
+  }
+}
+
+base::expected<BitcoinTransaction, std::string> KnapsackSolver::Solve() {
+  DCHECK_EQ(base_transaction_.inputs().size(), 0u);
+  DCHECK(base_transaction_.TargetOutput());
+  DCHECK(base_transaction_.ChangeOutput());
+
+  // TODO(apaymyshev): avoid dust inputs?
+  std::sort(input_groups_.begin(), input_groups_.end(), [](auto& g1, auto& g2) {
+    return g1.total_amount() > g2.total_amount();
+  });
+
+  std::multimap<uint64_t, BitcoinTransaction> solutions;
+
+  // Try to find best transaction with change output which receives fee
+  // surplus.
+  SolveForTransaction(base_transaction_, solutions);
+
+  // Drop change output from transaction and try to find best transaction
```

Comment:
nit: Drop the change output from the transaction and try to find the best transaction again.